src/lua/circllhist.lua

```diff
@@ -0,0 +1,565 @@
+        return f_lo / 0.75, 0
+      end
+    end
+  end
+end
+
+--------------------------------------------------------------------------------
+--- C struct based Histogram implementation
+Histogram_C.__index = Histogram_C
+
+function Histogram_C:__tostring()
+  local buf = string_builder:new()
+  local total = 0
+  for _, bin, count in self:funiter() do
+    buf:appendf("%.2g:%d", bin, count)
+    total = total + count
+  end
+  return string.format("Hist[%d]{%s}", total, buf:build(", "))
+end
+
+function Histogram_C:new(map)
+  local c_hist = libhist.hist_alloc()
+  assert(c_hist, "libhist.hist_alloc() returned nil (out of memory?!)")
+  ffi.gc(c_hist, libhist.hist_free)
+  if map then
+    for value, count in pairs(map) do
+      libhist.hist_insert(c_hist, value, count)
+    end
+  end
+  return c_hist
+end
+
+function Histogram_C:from_c_struct(c_hist, free_hist_when_done)
+  if c_hist == nil then
+    return ffi.cast("histogram_t*", nil)
+  end
+  if free_hist_when_done then
+    ffi.gc(c_hist, libhist.hist_free)
+  end
+  return c_hist
+end
+
+function Histogram_C:from_b64_string(b64_encoded_histo, str_len)
+  local c_hist = libhist.hist_alloc();
+  assert(c_hist, "libhist.hist_alloc() returned nil (out of memory?!)")
+  ffi.gc(c_hist, libhist.hist_free)
+
+  if str_len == nil then
+    str_len = string.len(b64_encoded_histo)
+  end
+
+  libhist.hist_deserialize_b64(c_hist, b64_encoded_histo, str_len)
+  return c_hist
+end
+
+function Histogram_C:bucket_count()
+  return libhist.hist_bucket_count(self)
+end
+
+function Histogram_C:clear()
+  libhist.hist_clear(self)
+end
+
+function Histogram_C:iter()
+  return coroutine.wrap(function()
+      for bin=0, self:bucket_count()-1 do
+        local value, count = self:bucket_idx(bin)
+        if count > 0 then
+          coroutine.yield(value, count)
+        end
+      end
+  end)
+end
+
+local function histogram_C_funiter_next(param, state)
+  local idx = state
+  if idx > param.limit then
+    return nil -- end iteration
+  end
+  local val, count = param.self:bucket_idx(idx)
+  return idx + 1, val, count
+end
+local function histogram_filter_count_zero(val, count)
+  return not (count == 0)
+end
+function Histogram_C:funiter()
+  local param = { self = self, limit = self:bucket_count()-1 }
+  local state = 0
+  return fun.iter(histogram_C_funiter_next, param, state)
+    :filter(histogram_filter_count_zero)
+end
+
+function Histogram_C:bucket_idx(bucket_id)
+  local value =  ffi.new("double[1]")
+  local count =  ffi.new("uint64_t[1]")
+  libhist.hist_bucket_idx(self, bucket_id, value, count)
+  return tonumber(value[0]), tonumber(count[0])
+end
+
+function Histogram_C:merge(other_histogram)
+  for _, b, c in other_histogram:funiter() do
+    self:insert(b,c);
+  end
+end
+
+function Histogram_C:merge_c(other_histogram_c)
+  local other_histo_ptr = ffi.new("const histogram_t*[1]", other_histogram_c)
+  libhist.hist_accumulate(self, other_histo_ptr, 1);
+end
```

Comment:
Why support two merge methods? This makes sense if we have multiple implementations for the same interface (which we had in Histogram_v1). There is only one here, so this is unmotivated.